File: 백준/Gold/11779. 최소비용 구하기 2/최소비용 구하기 2.cc
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
using namespace std;

struct Edge {
    int node;
    int cost;
    bool operator< (const Edge& a) const {
        return cost > a.cost;
    }
};
priority_queue<Edge> pq;
vector<Edge> adj[1001];
vector<Edge> rAdj[1001];
int dist[1001];
const int INF = 1e9;
int route[1001];

void dijkstra(int start, int end) {
    pq.push({start, 0});
    dist[start] = 0;
    while(!pq.empty()) {
        int node = pq.top().node;
        int cost = pq.top().cost;
        pq.pop();

        if (cost > dist[node]) continue;

        for (int i = 0; i < adj[node].size(); i++) {
            int next_node = adj[node][i].node;
            int next_cost = adj[node][i].cost;
            if (dist[next_node] > cost + next_cost) {
                route[next_node] = node;
                dist[next_node] = cost + next_cost;
                pq.push({next_node, dist[next_node]});
            }
        }
    }
}

void findRoute(int start, int end) {
    int node = end;
    stack<int> ans;
    while(node != start) {
        ans.push(node);
        node = route[node];
    }
    ans.push(start);
    cout << dist[end] << endl;
    cout << ans.size() << endl;
    while(!ans.empty()) {
        cout << ans.top() << " ";
        ans.pop();
    }
}
int main() {
    int n, m;
    cin >> n >> m;
    for (int i = 1; i <= n; i++) dist[i] = INF;
    for (int i = 0 ; i < m; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        adj[u].push_back({v, w});
        rAdj[v].push_back({u, w});
    }
    int start, end;
    cin >> start >> end;
    
    dijkstra(start, end);
    findRoute(start, end);

    return 0;
}

File: 백준/Gold/11779. 최소비용 구하기 2/최소비용 구하기 2.java
import java.io.*;
import java.util.*;

public class Main {

    static int INF = (int) 1e9;

    static int n, m;
    static List<Edge>[] graph;

    static class Edge {
        int to;
        int weight;

        public Edge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        m = Integer.parseInt(br.readLine());

        graph = new List[n + 1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            graph[a].add(new Edge(b, c));
        }

        StringTokenizer st = new StringTokenizer(br.readLine());
        int startNode = Integer.parseInt(st.nextToken());
        int endNode = Integer.parseInt(st.nextToken());

        // 역추적 및 다익스트라
        StringBuilder answer = new StringBuilder();
        int[] prev = new int[n + 1];

        int[] dist = new int[n + 1];
        dijkstra(startNode, endNode, dist, prev);

        answer.append(dist[endNode]).append("\n");

        Deque<Integer> stack = new LinkedList<>();
        for (int cur = endNode; cur != startNode; cur = prev[cur]) {
            stack.push(cur);
        }
        stack.push(startNode);

        answer.append(stack.size()).append("\n");
        while (!stack.isEmpty()) {
            answer.append(stack.pollFirst()).append(" ");
        }

        System.out.println(answer);
    }

    static void dijkstra(int start, int destination, int[] dist, int[] prev) {
        PriorityQueue<Edge> pq = new PriorityQueue<>((e1, e2) -> Integer.compare(e1.weight, e2.weight));
        Arrays.fill(dist, INF);
        pq.add(new Edge(start, 0));
        dist[start] = 0;

        while (!pq.isEmpty()) {
            Edge edge = pq.poll();
            int node = edge.to;
            int weight = edge.weight;

            if (dist[node] < weight) continue;

            for (Edge next : graph[node]) {
                int nextNode = next.to;
                int nextWeight = next.weight;

                if (dist[nextNode] > weight + nextWeight) {
                    dist[nextNode] = weight + nextWeight;
                    pq.add(new Edge(nextNode, dist[nextNode]));
                    prev[nextNode] = node;
                }
            }
        }
    }
}